# 01.09.23

# WORKING WITH DATASETS IN PANDAS

## Reading the Data

## Pandas can read a variety of files.
## eg, a table of fixed width formatted lines (read_fwf), excel sheets (read_excel), html files (read_html), json files (read_json), csv files (read_csv) etc.

## Syntax:
### pd.read_fileType(filepath)
### eg, pd.read_csv('auto_mpg.csv')

In [35]:
import pandas as pd
import numpy as np

# Creating a DateFrame
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/PYTHON FOR DS/0.FILES TO BE USED/auto_mpg.csv')
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa                 chevy s-10  

[398 rows x 9 columns]

# EXPLORING THE DATA

## 1.Head and Tail

### To view the first few rows or the last few rows, the functions that can be used are df.head() and df.tail respectively.
### It the no. of rows to be viewed is not passed, then, the head and tail funcitons provides five rows by default.

In [8]:
df.head(3)

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite

In [7]:
df.tail()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin             name  
393          82     usa  ford mustang gl  
394          82  europe        vw pickup  
395          82     usa    dodge rampage  
396          82     usa      ford ranger  
397          82     usa       chevy s-10

## 2.Describe

### The describe() fn can be used to generate a quick summary of data statistics.

In [9]:
df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

## 3. Info

### The info() fn is used to know about the datatypes and no. of rows containing null values for respective columns.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


## 4. Dropping Null Values

### It can be observed that the 'horsepower' attribute has some null values.
### The easiest approach is to remove the rows with any null values.
### This can be achieved using dropna() fn.

In [11]:
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model_year    392 non-null    int64  
 7   origin        392 non-null    object 
 8   name          392 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 30.6+ KB


#### After dropping the rows with null horsepower values, it can be observed that the no. of rows has been reduced to 392.

### NOTE:

*   'inplace' makes changes to the original DataFrame.
*   df.fillna(condition) can be used to fill all the missing values. The missing values are filled with mean, median, mode, or constant values.



## 5. Selecting a Subset of the Data

### In addition to data access techniques, pandas also provides techniques for indexing and selection.
### Selecting a specific column in a DataFrame can be achieved in following ways:

#### a. Passing the column name

In [13]:
df['name']
# O/p is a Series containing car names

0      chevrolet chevelle malibu
1              buick skylark 320
2             plymouth satellite
3                  amc rebel sst
4                    ford torino
                 ...            
393              ford mustang gl
394                    vw pickup
395                dodge rampage
396                  ford ranger
397                   chevy s-10
Name: name, Length: 392, dtype: object

#### b. Passing the column name as a list

In [36]:
df[['name']]
# O/p is a DataFrame containing just one column

name
0    chevrolet chevelle malibu
1            buick skylark 320
2           plymouth satellite
3                amc rebel sst
4                  ford torino
..                         ...
393            ford mustang gl
394                  vw pickup
395              dodge rampage
396                ford ranger
397                 chevy s-10

[398 rows x 1 columns]

No charts were generated by quickchart


#### c. Passing columns names in a list

In [26]:
df[['name', 'origin', 'model_year', 'mpg']]
# O/p is a DataFrame containing four columns

name  origin  model_year   mpg
0    chevrolet chevelle malibu     usa          70  18.0
1            buick skylark 320     usa          70  15.0
2           plymouth satellite     usa          70  18.0
3                amc rebel sst     usa          70  16.0
4                  ford torino     usa          70  17.0
..                         ...     ...         ...   ...
393            ford mustang gl     usa          82  27.0
394                  vw pickup  europe          82  44.0
395              dodge rampage     usa          82  32.0
396                ford ranger     usa          82  28.0
397                 chevy s-10     usa          82  31.0

[392 rows x 4 columns]

# INDEXERS

## Setting Custom Index

### Custom index can be set to the DataFrame according to the requirements.
### The set_index() fn is used.

In [17]:
# Example:

# creating a subset using head
df_head = df.head()

# setting name as custom index
df_head.set_index('name', inplace=True)
df_head

mpg  cylinders  displacement  horsepower  weight  \
name                                                                           
chevrolet chevelle malibu  18.0          8         307.0       130.0    3504   
buick skylark 320          15.0          8         350.0       165.0    3693   
plymouth satellite         18.0          8         318.0       150.0    3436   
amc rebel sst              16.0          8         304.0       150.0    3433   
ford torino                17.0          8         302.0       140.0    3449   

                           acceleration  model_year origin  
name                                                        
chevrolet chevelle malibu          12.0          70    usa  
buick skylark 320                  11.5          70    usa  
plymouth satellite                 11.0          70    usa  
amc rebel sst                      12.0          70    usa  
ford torino                        10.5          70    usa

### 'iloc' and 'loc' are the two indexing techniques that help us in selecting specific rows and columns.


In [19]:
# XYZ Custom Cars DataFrame (for upcoming code reference)
df

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa                 chevy s-10  

[392 rows x 9 columns]

#### 1. iloc: Accesses a group of rows and columns by integer index.
#### The 'iloc' indexer follows implicite index.
#### Syntax: df.iloc[rows,cols]

In [20]:
df.iloc[2,4]

3436

In [21]:
df.iloc[4,-1]

'ford torino'

In [24]:
df.iloc[1:5, 2:6]

displacement  horsepower  weight  acceleration
1         350.0       165.0    3693          11.5
2         318.0       150.0    3436          11.0
3         304.0       150.0    3433          12.0
4         302.0       140.0    3449          10.5

#### 2. loc: Accesses a group of rows and columns by custom index.
#### The 'loc' indexer follows explicite indexing.

#### NOTE: While retrieving records using loc, the upper range of slice is inclusive.
#### ie, range = [start,end]


In [25]:
# Creating a subset using head.
df_head = df.head()

# Setting 'name' as custom index
df_head.set_index('name', inplace=True)
df_head['buick skylark 320': 'ford torino']

mpg  cylinders  displacement  horsepower  weight  \
name                                                                    
buick skylark 320   15.0          8         350.0       165.0    3693   
plymouth satellite  18.0          8         318.0       150.0    3436   
amc rebel sst       16.0          8         304.0       150.0    3433   
ford torino         17.0          8         302.0       140.0    3449   

                    acceleration  model_year origin  
name                                                 
buick skylark 320           11.5          70    usa  
plymouth satellite          11.0          70    usa  
amc rebel sst               12.0          70    usa  
ford torino                 10.5          70    usa

### To select a subset of columns, the column names can be passed as a list.

In [28]:
df.loc[0:5, ['cylinders', 'horsepower', 'name']]

cylinders  horsepower                       name
0          8       130.0  chevrolet chevelle malibu
1          8       165.0          buick skylark 320
2          8       150.0         plymouth satellite
3          8       150.0              amc rebel sst
4          8       140.0                ford torino
5          8       198.0           ford galaxie 500

# ADDING OR REMOVING COLUMNS IN A DATEFRAME

In [29]:
# Creating a Python dictionary
marks = {'PR':[90,73],
         'DIP':[88,85],
         'BIC':[69,81]}
marks_df = pd.DataFrame(marks, index = ['Student1', 'Student2'])
marks_df

PR  DIP  BIC
Student1  90   88   69
Student2  73   85   81

In [33]:
# Adding a Column 'Total'
marks_df['Total'] = marks_df['PR'] + marks_df['DIP'] + marks_df['BIC']
marks_df

PR  DIP  BIC  Total
Student1  90   88   69    247
Student2  73   85   81    239

In [34]:
# Removing/Dropping the Column 'Total'
marks_df.drop(columns='Total', inplace=True)
marks_df

PR  DIP  BIC
Student1  90   88   69
Student2  73   85   81